pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,315 @@
+	kubeClient          client.Client
+	recorder            events.Recorder
+	cluster             *state.Cluster
+	clock               clock.Clock
+	provisioner         *provisioning.Provisioner
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate disruption commands
+func NewQueue(kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock,
+	provisioner *provisioning.Provisioner) *Queue {
+	queue := &Queue{
+		RateLimitingInterface: workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		ProviderIDToCommand:   map[string]*Command{},
+		kubeClient:            kubeClient,
+		recorder:              recorder,
+		cluster:               cluster,
+		clock:                 clock,
+		provisioner:           provisioner,
+	}
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, timeAdded time.Time, method string, consolidationType string) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates:        candidates,
+		Method:            method,
+		ConsolidationType: consolidationType,
+		TimeAdded:         timeAdded,
+	}
+}
+
+func (q *Queue) Name() string {
+	return "disruption.queue"
+}
+
+func (q *Queue) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+func (q *Queue) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	// The queue depth is the number of commands currently being considered.
+	// This should not reference the RateLimitingInterface, as there can be items that
+	// haven't been requeued yet due to their per-item backoff.
+	disruptionQueueDepthGauge.Set(float64(len(lo.Uniq(lo.Values(q.ProviderIDToCommand)))))
+	// Check if the queue is empty. client-go recommends not using this function to gate the subsequent
+	// get call, but since we're popping items off the queue synchronously, there should be no synchonization
+	// issues.
+	for q.Len() == 0 {
+		// time.Sleep(1 * time.Second)
+		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+	// Get command from queue. This waits until queue is non-empty.
+	item, shutdown := q.RateLimitingInterface.Get()
+	if shutdown {
+		panic("unexpected failure, disruption queue has shut down")
+	}
+	cmd := item.(*Command)
+	defer q.RateLimitingInterface.Done(cmd)
+
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if err := q.WaitOrTerminate(ctx, cmd); err != nil {
+		// If recoverable, re-queue and try again.
+		if !IsUnrecoverableError(err) {
+			// store the error that is causing us to fail so we can bubble it up later if this times out.
+			cmd.LastError = err
+			q.RateLimitingInterface.AddRateLimited(cmd)
```

Comment:
Oddly, this is what controller runtime is doing for reconcile.Requests for normal controllers. I wonder if there's a way for us to make this a standard controller, and drive it via a queue of keys.

Maybe something to explore in the future.

Reply:
Yeah, @jonathan-innis and I were talking about this. Could be something like a custom source that takes in events that are triggered by a queue.Add(). The difference is if the object we're reconciling (in this case it's a command) is in-memory vs in the cluster, where a custom source may be overkill (or may help in ways we're not aware of atm).